File: cpp/questions/isPalindrome1.cpp
#include <vector>

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        int ll = 0;
        ListNode* cn = head;
        while (cn) {
            cn = cn->next;
            ++ll;
        }
        int hl = ll / 2;
        std::vector<int> pl(hl);
        while (hl--) {
            pl[hl] = head->val;
            head = head->next;
        }
        head = ll%2 ? head->next : head;
        for (int i=0; i<pl.size(); ++i) {
            if (head->val != pl[i]) {
                return false;
            }
            head = head->next;
        }
        return true;
    }
};

File: python/questions/isPalindrome1.py
from typing import Optional

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next

class Solution:
    def isPalindrome(self, head: Optional[ListNode]) -> bool:
        ll = 0
        ch = head
        while ch:
            ch = ch.next
            ll += 1
        hl = ll // 2
        pl = [0] * hl
        while hl:
            hl -= 1
            pl[hl] = head.val
            head = head.next
        head = head.next if ll % 2 else head
        for i in range(len(pl)):
            if head.val != pl[i]:
                return False
            head = head.next
        return True
